File: baekjoon/python/7576/7576.py
from collections import deque
import sys
sys.stdin = open('7576_input.txt', 'r')


m, n = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]


def gr(graph):
    for i in graph:
        print(i)
    print()


def bfs(start):
    # 익은 토마토에서 시작
    q = deque()
    for riped in start:
        q.append(riped)
    while q:
        # gr(graph)
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            # 벽이면
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            # 안익은 토마토일때만 / 벽 일때, 익은 토마토일때 X
            if graph[nx][ny] == 0:
                # 익는 날짜 추가
                graph[nx][ny] = graph[x][y] + 1
                # 내일 익는 토마토 큐에 추가
                q.append((nx, ny))


fail = False


# 익은 토마토 위치 확인
pos = []
for i in range(n):
    for j in range(m):
        if graph[i][j] == 1:
            pos.append((i, j))
bfs(pos)

for lst in graph:
    if 0 in lst:
        fail = True
if fail:
    print(-1)
else:
    print(max(map(max, graph))-1)

File: baekjoon/cpp/7576/7576.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int n, m;

vector<int> graph[1001];
bool visited[1001][1001] = {
    false,
};

void bfs(queue<pair<int, int>> &q)
{
    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};
    int nx, ny;
    while (!q.empty())
    {
        int x, y;
        x = q.front().first;
        y = q.front().second;
        q.pop();
        for (int i = 0; i < 4; i++)
        {
            nx = x + dx[i];
            ny = y + dy[i];
            if (0 <= nx && nx < n && 0 <= ny && ny < m && !visited[nx][ny] && graph[nx][ny] != -1)
            {
                if (!graph[nx][ny])
                {
                    visited[nx][ny] = true;
                    graph[nx][ny] = graph[x][y] + 1;
                    q.push({nx, ny});
                }
            }
        }
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    freopen("7576_input.txt", "r", stdin);

    cin >> m >> n;
    int tmp;
    queue<pair<int, int>> tom_q;
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            cin >> tmp;
            graph[i].push_back(tmp);
            if (tmp == 1)
            {
                tom_q.push({i, j});
                visited[i][j] = true;
            }
        }
    }

    // while (!tom_q.empty())
    // {
    //     cout << tom_q.front().first << "," << tom_q.front().second << endl;
    //     tom_q.pop();
    // }

    // for (int i = 0; i < n; i++)
    // {
    //     for (const auto &elem : graph[i])
    //     {
    //         cout << elem << " ";
    //     }
    //     cout << endl;
    // }

    bfs(tom_q);

    int max = -1;
    for (int i = 0; i < n; i++)
    {
        if (find(graph[i].begin(), graph[i].end(), 0) != graph[i].end())
        {
            cout << -1 << endl;
            return 0;
        }
        int linemax = *max_element(graph[i].begin(), graph[i].end());

        if (linemax > max)
            max = linemax;

        // for (const auto &elem : graph[i])
        // {
        //     cout << elem << " ";
        // }
        // cout << endl;
    }
    cout << max - 1 << endl;
    // cout << graph[n - 1][m - 1] << endl;
    return 0;
}
